Write secret files atomically

szh opened the pull request:
### Desired Outcome

This functionality is described here:

https://github.com/cyberark/secrets-provider-for-k8s/blob/secrets-rotation-design/design/secret_rotation_design.md#atomic-file-writes-for-push-to-file-mode


### Implemented Changes

Files are written to a temp file first. If that succeeds, the temp file is moved to the destination in an atomic operation.

### Connected Issue/Story

CyberArk internal issue link: [ONYX-16710](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-16710)

### Definition of Done

- [x] Investigate if it's possible to test atomicity
- [x] Use atomic operation to write secrets files
- [x] Address security questions raised in #428 ([this](https://github.com/cyberark/secrets-provider-for-k8s/pull/428#discussion_r794972112) and [this](https://github.com/cyberark/secrets-provider-for-k8s/pull/428#discussion_r794975481) comment)

#### Changelog

- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [x] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [x] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

doodlesbykumbi commented on NOTICES.txt:
```diff
@@ -34,6 +34,7 @@ Section 3: MIT
 >>> github.com/stretchr/testify - v1.7.0
 >>> gopkg.in/yaml.v3 - v3.0.0-20210107192922-496545a6307b
 >>> sigs.k8s.io/yaml - v1.3.0
+>>> github.com/natefinch/atomic - v1.0.1
```
I think you previously mentioned the idea of creating a fork of this package (though I forget the reason). I'm wondering if it might be worth just adding an atomic file writer to a package in this repo. The implementation is on the small side. The atomic write implementations I've seen elsewhere write to a temp file then use `os.Rename` to finish the job.

szh replied:
The reason for the fork was because the original package doesn't report if the temp file couldn't be deleted. Here's what I'm looking to add: https://github.com/natefinch/atomic/pull/17
It is indeed a small feature to implement, just a couple of go files. The only complicated part of it is the [Windows](https://github.com/natefinch/atomic/blob/master/file_windows.go) implementation, but I'm not sure we even need that.

tzheleznyak commented on pkg/atomicwriter/atomic_writer.go:
```diff
@@ -0,0 +1,55 @@
+		// The temporary file could not be deleted by the 'atomic' package.
+		// Try to delete the temporary file again asyncronously.
+		go func() {
+			time.Sleep(time.Second * 5)
```
Move 5 to const

doodlesbykumbi commented on pkg/atomicwriter/atomic_writer.go:
```diff
@@ -0,0 +1,108 @@
+// takes the file path during construction, so the code which calls
+// `Write()` doesn't need to be concerned with the destination, just like
+// any other writer.
+type AtomicWriter struct {
+	path        string
+	permissions os.FileMode
+	tempFile    *os.File
+}
+
+func NewAtomicWriter(path string, permissions os.FileMode) (*AtomicWriter, error) {
```
For the sake of limiting what is made public to that which is necessary I think we should make the struct definition private `AtomicWriter => atomicWriter` and `NewAtomicWriter` can return an interface`(io.WriteCloser, error)`.

szh replied:
Changed in 3e1def1ede9eacd3d5d24d268036dd2770aaddaa

doodlesbykumbi commented on pkg/atomicwriter/atomic_writer.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+func (w *AtomicWriter) Close() error {
+	defer func() {
+		go w.Cleanup()
+	}()
```
1. Why does this need retry logic ?
2. Deferring a call to a go routine feels like trouble

szh replied:
1. See [this thread](https://github.com/cyberark/secrets-provider-for-k8s/pull/428#discussion_r794975481)
2. Why? Is there a preferred way to do this? My assumption is that a regular defer() function will block execution, which is a problem if we're doing timed retries.

doodlesbykumbi replied:
1. I'm not sure about the appropriateness of retry here. I think this is an instance of an application assuming the reliability responsibilities of something like the OS/file system. I think we ought to be able to safely assume reliability here. I understand that deletion might fail, but I think a single attempt + errors should be enough. An alternative might be to try and delete and if it doesn't work Truncate the file to 0, which is highly likely to work if done after successful writes. 
2. It's sort of a code smell, at least in my eyes. Defer creating Go routines that are not tracked by anything... An alternative, if we really decide to keep an eye on those tmp file and make sure they are cleaned up might be to a have some sort of temp file manager that keeps track of a collection (maybe a queue) of temp files and in the background in a single goroutine shared between all the files can carry out the regular task of cleaning up the temp files.

szh replied:
@andytinkham What do you think?

andytinkham replied:
It feels like truncate to 0 solves the problem of possibly leaking secrets which is my main concern. If that's easier, I'm ok with it.  For #2, I don't think I have any specific thoughts - seems OK.

szh replied:
OK. So instead of the whole retry logic, we just
1. Attempt to delete the file
2. If that fails, attempt to truncate it (and maybe log it)
3. If that fails, log the error and exit
Is that right?

diverdane reviewed (commented):
@szh I've added some comments re. one way to mock the `os.Remove` function for testing to your query on the `about-golang` Teams channel (along with the `os.Chmod` and `os.Rename` functions).

szh commented on pkg/atomicwriter/atomic_writer.go:
```diff
@@ -0,0 +1,109 @@
+		dir = "."
+	}
+
+	f, err := ioutil.TempFile(dir, file)
```
@andytinkham I'm thinking it would be better to put this in a temp directory so it's not in the shared volume. Do you agree?

szh replied:
Update: putting this in a temp folder results in a `invalid cross-device link` error when trying to call os.rename when the destination directory is on a different volume.

szh commented:
> @szh I've added some comments re. one way to mock the `os.Remove` function for testing to your query on the `about-golang` Teams channel (along with the `os.Chmod` and `os.Rename` functions).

Thanks! I've added some unit tests the ideas you posted.

diverdane commented on pkg/atomicwriter/atomic_writer_test.go:
```diff
@@ -0,0 +1,243 @@
+				// Check the file permissions
+				mode, err := os.Stat(path)
+				assert.NoError(t, err)
+				assert.Equal(t, os.FileMode(0644), mode.Mode())
```
Coverage looks really good!

For completeness... would it be possible to check that the temp file is successfully deleted in the happy path? (Or maybe I'm missing where that's tested already?) I'm not sure if it's easier to add here (maybe return the `tempFile` in your `writeFile()` function, so it can be checked by the caller?), or some other happy path case.

diverdane reviewed (commented):
Looks terrific!
One minor comment/question, otherwise good-to-go!!!

szh commented:
Requesting security review from @andytinkham

diverdane reviewed (approved):
LGTM!!!

andytinkham reviewed (approved):
LGTM!

doodlesbykumbi commented on pkg/atomicwriter/atomic_writer_test.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+// Mocks the logger output to a buffer
+func mockErrorLog() *bytes.Buffer {
+	buf := &bytes.Buffer{}
+	log.ErrorLogger.SetOutput(buf)
+	return buf
+}
+
+// Unmocks the logger output, sending it back to os.Stderr
+func unmockErrorLog() {
+	log.ErrorLogger.SetOutput(os.Stderr)
+}
```
In this case it doesn't seem that we have to worry about running tests in parallel but if we did I suspect the mocking and un-mocking might breakdown. To make testing the logs simpler I would have dependency injected the error logger function, much like you did the os funcs, since the atomic writer only uses `log.Error`.

doodlesbykumbi commented on pkg/secrets/provide_conjur_secrets_test.go:
```diff
@@ -143,7 +144,7 @@ func TestSecretProvider(t *testing.T) {
 			mode:          "sidecar",
 			interval:      time.Duration(10) * time.Millisecond,
-			testTime:      time.Duration(65) * time.Millisecond,
+			testTime:      time.Duration(65+atomicWriteDelayMsecs) * time.Millisecond,
```
What's all this about ? (Just curious)

szh replied:
In the case where we're doing atomic writes, it adds a very small delay which causes the retry testing to get thrown off (though only on Jenkins, not when running locally). This fixes that.

doodlesbykumbi reviewed (commented):
Left a couple comments. Otherwise PR looking good
